zebra-state/src/service/non_finalized_state.rs

```diff
@@ -35,23 +35,71 @@ pub(crate) use chain::Chain;
 /// which returns a shared reference to the database.
 ///
 /// Most chain data is clone-on-write using [`Arc`].
-#[derive(Clone, Debug)]
 pub struct NonFinalizedState {
-    /// Verified, non-finalized chains, in ascending order.
+    /// Verified, non-finalized chains, in ascending work order.
     ///
-    /// The best chain is `chain_iter().next()`.
-    /// Using `chain_set.last()` or `chain_set.iter().next_back()` is deprecated, and should migrate to `chain_iter().next()`.
+    /// The best chain is [`NonFinalizedState::best_chain()`], or `chain_iter().next()`.
+    /// Using `chain_set.last()` or `chain_set.iter().next_back()` is deprecated,
+    /// callers should migrate to `chain_iter().next()`.
     chain_set: BTreeSet<Arc<Chain>>,
 
     /// The configured Zcash network.
     pub network: Network,
 
-    #[cfg(feature = "getblocktemplate-rpcs")]
+    // Diagnostics
+    //
     /// Configures the non-finalized state to count metrics.
     ///
-    /// Used for skipping metrics counting when testing block proposals
+    /// Used for skipping metrics and progress bars when testing block proposals
     /// with a commit to a cloned non-finalized state.
-    pub should_count_metrics: bool,
+    //
+    // TODO: make this field private and set it via an argument to NonFinalizedState::new()
+    #[cfg(feature = "getblocktemplate-rpcs")]
+    should_count_metrics: bool,
+
+    /// Number of chain forks transmitter.
+    #[cfg(feature = "progress-bar")]
+    chain_count_bar: Option<howudoin::Tx>,
+
+    /// A chain fork length transmitter for each [`Chain`] in [`chain_set`](Self.chain_set).
+    ///
+    /// Because `chain_set` contains `Arc<Chain>`s, it is difficult to update the metrics state
+    /// on each chain. ([`Arc`]s are read-only, and we don't want to clone them just for metrics.)
+    #[cfg(feature = "progress-bar")]
+    chain_fork_length_bars: Vec<howudoin::Tx>,
+}
+
+impl std::fmt::Debug for NonFinalizedState {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut f = f.debug_struct("NonFinalizedState");
+
+        f.field("chain_set", &self.chain_set)
+            .field("network", &self.network);
+
+        #[cfg(feature = "getblocktemplate-rpcs")]
+        f.field("should_count_metrics", &self.should_count_metrics);
+
+        f.finish()
+    }
+}
+
+impl Clone for NonFinalizedState {
+    fn clone(&self) -> Self {
+        Self {
+            chain_set: self.chain_set.clone(),
+            network: self.network,
+
+            #[cfg(feature = "getblocktemplate-rpcs")]
+            should_count_metrics: self.should_count_metrics,
```

Comment:
Optional/Nitpick: Not strictly related to this PR but could we set `should_count_metrics` to false when cloning to replace the call to `disable_metrics()` in the `CheckBlockProposalValidity` request?

```suggestion
            should_count_metrics: false,
```

Reply:
Good idea, I think that would be a bit more robust if we add future requests that clone the state.

Did you want to open a cleanup PR on top of this one?
(I need to merge this PR so I can finish the blog post and move on with other work.)